### ISSUE TITLE
Dev Requests Form: JS alert shows when no crumb is found in cookies

### ISSUE THREAD

**Issue by retzion** [brave/brave-website#29]
![pasted image at 2016_01_27 11_51 pm](https://cloud.githubusercontent.com/assets/1813451/12652085/8b65840a-c5a6-11e5-8b0c-d77a5199db05.png)

We show an alert upon the form post if we don't find the cookie named "crumb". The cookie is required for all posts to the server.

Perhaps we can insert a method here to ping our api server for a new crumb. Upon success, we repost the form. We can set a number of retries and only then show some kind of alert.

**Comment by diracdeltas** [comment]
that approach sounds good. this cookie is needed for csrf protection. our server should be setting it for every user when they visit the page. not sure why it isn't

**Comment by bbondy** [comment]
I think related but someone on twitter is reporting that our browser doesn't support our own site.
<img width="1025" alt="screenshot 2016-01-30 21 07 35" src="https://cloud.githubusercontent.com/assets/831718/12699755/9a307ff8-c795-11e5-9833-8194a0622972.png">

Why is this labeled wontfix?

**Comment by AlfonsoML** [comment]
Yep, I posted that screenshot. 
It failed in all the browsers and debugging the code I found that indeed it was trying to read a cookie and it was empty although I don't block cookies.
So I disabled uBlock origin and after reloading the page I was able to submit the form so I guess that the cookie is somehow set by a 3rd party script or anything else that it's commonly blocked by ad blockers, and you should review the site so the csrf protection still exists, but it's done in a way compatible with your own browser.

**Comment by diracdeltas** [comment]
The first-party cookie (`crumb`) is actually getting set correctly on the staging server but not in prod. Weird that disabling uBlock fixed it, since it shouldn't be third-party. will look into it

**Comment by diracdeltas** [comment]
i don't see difference in this repo between staging and master that could cause this problem. also, i don't see any set-cookie headers on a fresh browser profile being sent from the staging server so i assume it's getting set by page javascript. maybe this is related to heroku/fastly? cc @bridiver

**Comment by diracdeltas** [comment]
fixed on live site now

**Comment by retzion** [comment]
@ElissaBeth tells me the fix is not working? I can work on it tonight.
I plan to add and use a recursive function on lines 14 and 60 here https://github.com/brave/brave-website/blob/staging/public/assets/js/forms/mailchimp.js
The function will retrieve the cookie from the client or keep trying X times via an ajax post to the server.

**Comment by diracdeltas** [comment]
@ElissaBeth please give steps to reproduce the not-working part. i signed up successfully just now.

**Comment by ElissaBeth** [comment]
I have not reproduced the issue. It is intermittent.

On Wed, Feb 3, 2016 at 12:40 PM, yan zhu (@bcrypt) <notifications@github.com
> wrote:

> @ElissaBeth <https://github.com/ElissaBeth> please give steps to
> reproduce the not-working part. i signed up successfully just now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/brave/brave-website/issues/29#issuecomment-179453342>.
>



-- 
Elissa Shevinsky
Twitter.com/ElissaBeth
650-380-6972

**Comment by retzion** [comment]
@ElissaBeth I implemented the fix I mentioned above on [staging](https://evarb.brave.com). Everything seems to work but I'm also not able to replicate the error with the cookie. In my testing I manually deleted the crumb cookie upon form post and the recursive method successfully retrieved a new crumb from the api. I also removed the js alert we were showing upon failure. The form post currently retries for a crumb 3 times then it will just hang. I'm not sure how you want to handle this. I know there was talk about not showing an alert.
Also, I have only implemented this on the Build Requests form, not the Newsletter Subscription form in the footer. I'll get to that tomorrow. 
@diracdeltas please review my work as time permits and merge to master when you want to try it on the production site. (fingers crossed)
